File: 11-container-with-most-water/container-with-most-water.ts
/**
너비: (e - s)
높이: Math.min(height[s], height[e])

넓이: (e - s) * Math.min(height[s], height[e])
 */


function maxArea(height: number[]): number {
    let maxArea = 0;
    let s = 0, e = height.length - 1;
    while (s < e) {
        maxArea = Math.max((e - s) * Math.min(height[s], height[e]), maxArea)
        if (height[s] > height[e]) {
            e -= 1
        } else {
            s += 1
        }
    }
    return maxArea;
};

File: 11-container-with-most-water/container-with-most-water.py
"""
너비: e - s
높이: min(height[s], height[e])

넓이 = e - s * min(height[s], height[e])
"""

class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_area = 0
        s, e = 0, len(height) - 1
        while s < e:
            area = (e - s) * min(height[s], height[e])
            max_area = max(area, max_area)
            if height[s] < height[e]:
                s += 1
            else:
                e -= 1

        return max_area
